packages/api/src/sequelize/migrations/2023-07-25_01_create-docref-mapping.ts

```diff
@@ -0,0 +1,53 @@
+import { DataTypes } from "sequelize";
+import type { Migration } from "..";
+import * as shared from "../migrations-shared";
+
+const tableName = "docref_mapping";
+
+// Use 'Promise.all' when changes are independent of each other
+export const up: Migration = async ({ context: queryInterface }) => {
+  await queryInterface.sequelize.transaction(async transaction => {
+    await shared.createTable(
+      queryInterface,
+      tableName,
+      {
+        id: {
+          type: DataTypes.STRING,
+          primaryKey: true,
+        },
+        externalId: {
+          type: DataTypes.STRING,
+          field: "external_id",
+        },
+        cxId: {
+          type: DataTypes.UUID,
+          field: "cx_id",
+        },
+        patientId: {
+          type: DataTypes.STRING,
+          field: "patient_id",
+        },
+        source: {
+          type: DataTypes.STRING,
+        },
+      },
+      {
+        transaction,
+        addVersion: true,
+        uniqueKeys: {
+          external_unique: { fields: ["external_id", "patient_id", "cx_id", "source"] },
+        },
+      }
+    );
+    await queryInterface.addIndex(tableName, ["cx_id", "patient_id", "source", "external_id"], {
```

Comment:
nit: can define `["external_id", "patient_id", "cx_id", "source"]` in a const and use here and above

Reply:
No, the order makes all the difference when building indexes, and each have its own order of elements.

Reply:
Cool, didn't know this was intentional

Reply:
Yeah, the order of columns on the index is critical:
- the DB will traverse the index from "left to right", so in this case it will start with `cx_id`;
- ideally, we would start with the column that results in least records after we filter by the first column, so `cx_id` is not the first best candidate at all (the best one would be `id`, but that's already an index b/c of PK and it wouldn't solve the next item);
- because of the traversing order, this index can't be used if we wanted to query only by `patient_id` (if we do so with the indexes created after this migration, it would result in a full table scan);
- but, because we should always include `cx_id` on queries, making `cx_id` the first column allow us to filter all records of a given customer, in addition to also filter by patient;
- same applies to the third column.

Wanted to do a larger write up so we can refer to it in case someone's wondering how DB indexes work.